tpm2/test/cipher/aes_gcm_test.c

```diff
@@ -0,0 +1,307 @@
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 0);
+
+  free(decrypt);
+  decrypt = NULL;
+  decrypt_len = 0;
+
+  // modify a single key bit
+  key[0] ^= 1;
+
+  // verify key modification breaks decryption
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 1);
+
+  free(plaintext);
+  free(ciphertext);
+  free(key);
+}
+	    
+void test_gcm_tag_modification(void)
+{
+  unsigned char* key        = NULL;
+  unsigned char* plaintext  = NULL;
+  unsigned char* ciphertext = NULL;
+  unsigned char* decrypt    = NULL;
+
+  int key_len = 16;
+  size_t plaintext_len = 16;
+  size_t ciphertext_len = 0;
+  size_t decrypt_len = 0;
+  
+  key = calloc(key_len, 1);
+  plaintext = calloc(plaintext_len, 1);
+
+  // check encryption/decryption of base case
+  aes_gcm_encrypt(key, key_len, plaintext, plaintext_len,
+                  &ciphertext, &ciphertext_len);
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 0);
+
+  free(decrypt);
+  decrypt = NULL;
+  decrypt_len = 0;
+
+  // alter the last byte of the tag and verify decryption failure
+  ciphertext[ciphertext_len-1] ^= 0x1;
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 1);
+  decrypt_len = 0;
+  decrypt = NULL;
+ 
+  // truncate the tag by 2 bytes and verify decryption failure
+  ciphertext_len -= 2;
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 1);
+  decrypt_len = 0;
+  decrypt = NULL;
+  
+  free(key);
+  free(plaintext);
+  free(ciphertext);
+}
+
+void test_gcm_iv_modification(void)
+{
+  unsigned char* key        = NULL;
+  unsigned char* plaintext  = NULL;
+  unsigned char* ciphertext = NULL;
+  unsigned char* decrypt    = NULL;
+
+  int key_len = 16;
+  size_t plaintext_len = 16;
+  size_t ciphertext_len = 0;
+  size_t decrypt_len = 0;
+  
+  key = calloc(key_len, 1);
+  plaintext = calloc(plaintext_len, 1);
+
+  // check encryption/decryption of base case
+  aes_gcm_encrypt(key, key_len, plaintext, plaintext_len,
+                  &ciphertext, &ciphertext_len);
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 0);
+
+  free(decrypt);
+  decrypt = NULL;
+  decrypt_len = 0;
+
+  // alter the first byte of IV and verify decryption failure
+  ciphertext[0] ^= 0x1;
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 1);
+  decrypt = NULL;
+  decrypt_len = 0;
+
+  // truncate the IV and verify decryption failure
+  unsigned char* truncated_iv_cipher = ciphertext + 2;
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, truncated_iv_cipher, ciphertext_len-2,
```

Comment:
Similar to above, won't this one also already fail without the iv modification because:

    ciphertext[0] ^= 0x1;

Reply:
Again, the modified byte (first byte ciphertext array is first byte of IV) is one that is not passed to aes_gcm_decrypt() in the truncated IV case (ciphertext array starts at byte two past the real first byte). Again, though, I think it made sense to reverse the order here to make the two cases more independent.